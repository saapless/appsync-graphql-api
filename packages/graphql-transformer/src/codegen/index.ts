export * from "./ast";
export * from "./printer";
export * from "./document";
