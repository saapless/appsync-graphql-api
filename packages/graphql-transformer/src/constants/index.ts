export const ConnectionDirective = Object.freeze({
  HAS_ONE: "hasOne",
  HAS_MANY: "hasMany",
});

export const ModelDirective = Object.freeze({
  MODEL: "model",
});

export const UtilityDirective = Object.freeze({
  SERVER_ONLY: "serverOnly",
  CLIENT_ONLY: "clientOnly",
  READ_ONLY: "readOnly",
  WRITE_ONLY: "writeOnly",
  FILTER_ONLY: "filterOnly",
  INTERNAL: "internal",
});

export const ScalarType = Object.freeze({
  STRING: "String",
  INT: "Int",
  FLOAT: "Float",
  BOOLEAN: "Boolean",
  ID: "ID",
  LONG: "Long",
});

export const RelationType = Object.freeze({
  ONE_TO_ONE: "oneToOne",
  ONE_TO_MANY: "oneToMany",
  MANY_TO_MANY: "manyToMany",
});
