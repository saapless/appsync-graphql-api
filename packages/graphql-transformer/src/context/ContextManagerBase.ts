import { TransformerContext } from "./TransformerContext";

export abstract class ContextManagerBase {
  protected readonly _context: TransformerContext;

  constructor(context: TransformerContext) {
    this._context = context;
  }
}
