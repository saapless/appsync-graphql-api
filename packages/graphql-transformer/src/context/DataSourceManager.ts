import { DynamoDbGenerator, NoneGenerator, ResolverGeneratorBase } from "../generators";
import { TransformExecutionError } from "../utils/errors";
import { TransformerContext } from "./TransformerContext";

export type NoneDataSorce = {
  type: "NONE";
};

export type DynamoDBDataSource = {
  type: "DYNAMO_DB";
  config: {
    tableName: string;
  };
};

export type DataSourceConfig = NoneDataSorce | DynamoDBDataSource;

export interface DataSourceManagerConfig {
  primaryDataSourceName: string;
  dataSources: Record<string, DataSourceConfig>;
}

export class DataSourceManager {
  private readonly _primaryDataSourceName: string;
  private readonly _dataSources: Map<string, DataSourceConfig>;
  private readonly _context: TransformerContext;

  constructor(context: TransformerContext, config: DataSourceManagerConfig) {
    this._context = context;
    this._primaryDataSourceName = config.primaryDataSourceName;
    this._dataSources = new Map(Object.entries(config.dataSources));

    this._validatePrimaryDataSourceConfig();
  }

  private _validatePrimaryDataSourceConfig() {
    const primaryDataSource = this._dataSources.get(this._primaryDataSourceName);

    if (!primaryDataSource || primaryDataSource.type !== "DYNAMO_DB") {
      throw new TransformExecutionError(
        "`defaultDataSourceName` needs to be configured as a `DYNAMO_DB` dataSource"
      );
    }
  }

  public get primaryDataSourceName(): string {
    return this._primaryDataSourceName;
  }

  public get noneDataSourceName(): string {
    const noneDataSource = Array.from(this._dataSources.entries()).find(
      (ds) => ds[1].type === "NONE"
    );

    if (!noneDataSource) {
      throw new TransformExecutionError("No `NONE` dataSource configured");
    }

    return noneDataSource[0];
  }

  public getDataSource(name: string) {
    const dataSource = this._dataSources.get(name);

    if (!dataSource) {
      throw new TransformExecutionError(`Missing dataSource ${name}`);
    }

    return dataSource;
  }

  public getDataSourceGenerator(dataSourceName: string): ResolverGeneratorBase {
    const dataSource = this.getDataSource(dataSourceName);

    switch (dataSource.type) {
      case "DYNAMO_DB":
        return new DynamoDbGenerator(this._context);
      case "NONE":
        return new NoneGenerator(this._context);
      default:
        throw new TransformExecutionError(
          `Generator not implemented for data source type: ${dataSource}`
        );
    }
  }
}
