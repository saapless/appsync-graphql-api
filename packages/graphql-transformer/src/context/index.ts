export * from "./TransformerContext";
export type {
  DescriptorOperation,
  Key,
  KeyOperator,
  KeyValue,
  OperationActionType,
  ResolverDescriptor,
  ReturnType,
} from "./ResolverLoader";
