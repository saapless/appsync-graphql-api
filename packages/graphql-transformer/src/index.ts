export * from "./transformer";
export { definitionFromSource } from "./utils";
