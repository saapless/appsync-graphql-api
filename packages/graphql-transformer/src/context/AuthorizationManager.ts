import { AuthorizationRule, ReadOperation, WriteOperation } from "../utils/types";
import { ContextManagerBase } from "./ContextManagerBase";
import { TransformerContext } from "./TransformerContext";

type AuthorizationMode = "API_KEY" | "USER_POOL" | "IAM" | "OIDC" | "LAMBDA";

export interface AuthorizationManagerConfig {
  additionalauthorizationModes?: AuthorizationMode[];
  defaultAuthorizationMode?: AuthorizationMode;
  defaultAuthorizationRules?: AuthorizationRule[];
}

export class AuthorizationManager extends ContextManagerBase {
  private readonly _defaultAuthorizationMode: AuthorizationMode;
  private readonly _additionalAuthorizationModes: AuthorizationMode[];
  private readonly _defaultAuthorizationRules: AuthorizationRule[];
  private readonly _modelAuthRules: Map<string, AuthorizationRule[]> = new Map();

  constructor(context: TransformerContext, config: AuthorizationManagerConfig) {
    super(context);

    this._defaultAuthorizationMode = config.defaultAuthorizationMode = "API_KEY";
    this._defaultAuthorizationRules = config.defaultAuthorizationRules ?? [];
    this._additionalAuthorizationModes = config.additionalauthorizationModes ?? [];
    this._modelAuthRules = new Map();
  }

  get defaultAuthRules() {
    return this._defaultAuthorizationRules;
  }

  public setModelRules(model: string, rules: AuthorizationRule[]) {
    this._modelAuthRules.set(model, rules);
  }

  public getAuthRules(
    operation: ReadOperation | WriteOperation,
    targetType: string,
    definedRules?: AuthorizationRule[]
  ) {
    if (definedRules) {
      return definedRules;
    }

    const modelRules = this._modelAuthRules.get(targetType) ?? this._defaultAuthorizationRules;

    const operationRules = modelRules.filter(
      (rule) => !rule.operations || rule.operations.includes(operation)
    );

    return operationRules;
  }
}
