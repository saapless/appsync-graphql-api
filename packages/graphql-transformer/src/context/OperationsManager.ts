import { Operation, ReadOperation, WriteOperation } from "../utils";
import { ContextManagerBase } from "./ContextManagerBase";
import { TransformerContext } from "./TransformerContext";

export interface OperationsManagerConfig {
  readOperations?: ReadOperation[];
  writeOperations?: WriteOperation[];
  defaultModelOperations?: Operation[];
}

export const DEFAULT_READ_OPERATIONS: ReadOperation[] = ["get", "list"];
export const DEFAULT_WRITE_OPERATIONS: WriteOperation[] = ["create", "update", "delete"];

export class OperationsManager extends ContextManagerBase {
  private readonly _readOperations: ReadOperation[];
  private readonly _writeOperations: WriteOperation[];
  private readonly _defaultModelOperations: (ReadOperation | WriteOperation)[];

  constructor(context: TransformerContext, config: OperationsManagerConfig = {}) {
    super(context);

    this._readOperations = config.readOperations ?? DEFAULT_READ_OPERATIONS;
    this._writeOperations = config.writeOperations ?? DEFAULT_WRITE_OPERATIONS;
    this._defaultModelOperations = this._expandOperations(config.defaultModelOperations);
  }

  private _expandOperations(operations?: Operation[]): (ReadOperation | WriteOperation)[] {
    if (!operations) {
      return [...DEFAULT_READ_OPERATIONS, ...DEFAULT_WRITE_OPERATIONS];
    }

    const expandedOperations: Set<ReadOperation | WriteOperation> = new Set();

    for (const operation of operations) {
      switch (operation) {
        case "read":
          this._readOperations.forEach((op) => expandedOperations.add(op));
          break;
        case "write":
          this._writeOperations.forEach((op) => expandedOperations.add(op));
          break;
        default:
          expandedOperations.add(operation);
      }
    }

    return Array.from(expandedOperations.values());
  }

  public getModelOperations(args?: Operation[]) {
    if (args?.length) {
      return this._expandOperations(args);
    }

    return this._defaultModelOperations;
  }
}
