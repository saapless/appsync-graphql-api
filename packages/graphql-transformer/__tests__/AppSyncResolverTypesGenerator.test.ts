import { jest } from "@jest/globals";
import { TransformerContext } from "../src/context";
import { DocumentNode } from "../src/definition";

let outputContent = "";
jest.unstable_mockModule("node:fs", () => {
  return {
    default: {
      writeFileSync: jest.fn().mockImplementation((path, content) => {
        outputContent = content as string;
      }),
    },
  };
});

const { AppSyncResolverTypesGenerator } = await import(
  "../src/generators/AppSyncResolverTypesGenerator"
);

const context = new TransformerContext({
  outputDirectory: "__test__",
  document: DocumentNode.fromSource(/* GraphQL */ `
    type User {
      id: ID!
      name: String!
    }

    input CreateUserInput {
      id: ID
      name: String
    }

    input UpdateUserInput {
      id: ID!
      name: String
    }

    type Query {
      me: User
    }

    type Mutation {
      createUser(input: CreateUserInput!): User
      updateUser(input: UpdateUserInput!): User
      deleteUser(id: ID!): User
    }
  `),
});

describe("ResolverTypesGenerator", () => {
  beforeAll(() => {
    context.resolvers.setLoader("Query", "me", {
      targetName: "User",
      action: {
        type: "getItem",
        key: {
          id: { ref: "identity.sub" },
        },
      },
      returnType: "result",
    });
    context.resolvers.setLoader("Mutation", "createUser", {
      targetName: "User",
      action: {
        type: "putItem",
        key: { id: { ref: "identity.sub" } },
      },
      returnType: "result",
    });
    context.resolvers.setLoader("Mutation", "updateUser", {
      targetName: "User",
      action: {
        type: "updateItem",
        key: { id: { ref: "identity.sub" } },
      },
      returnType: "result",
    });
    context.resolvers.setLoader("Mutation", "deleteUser", {
      targetName: "User",
      action: {
        type: "removeItem",
        key: { id: { ref: "args.id" } },
      },
      returnType: "result",
    });
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it("generates resolver context types", async () => {
    const generator = new AppSyncResolverTypesGenerator(context);

    generator.generate("__test__");
    expect(outputContent).toMatchSnapshot();
  });
});
