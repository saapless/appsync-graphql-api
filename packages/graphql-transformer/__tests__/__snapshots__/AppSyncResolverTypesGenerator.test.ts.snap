// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolverTypesGenerator generates resolver context types 1`] = `""`;
